### PR TITLE
Back to Fix Reporting Issues OA Limiting Factor

### LINKED ISSUE #8802: Air System Outdoor Air Limiting Factor can be wrong for simple OA controls due to tiny differences
Issue overview
--------------
Air System Outdoor Air Limiting Factor uses several < or > tests to determine if the OA flow rate has been constrained by a particular limiting factor (e.g. economizer, OA schedule, etc.). For simple OA controllers with no economizer, no min OA schedule, and no other constraints, this should always be zero. But occasional 1s show up due to tiny differences in comparing two values which should be equal, but due to different calculation paths for the two values, they differ out in the 10th decimal or so. Probably need to add a tolerance to one or more of these comparisons, and review the entire logic for this output.

Also, this output is reset to zero every time step inside `CalcOAController` which I'm pretty sure is not called if an airloop is off. Need a test file to prove this.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 9.6.0-b8f984adb5
 - Discovered while investigating unexpected diffs in #8759 

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #8802
 - Fix loop flow issues when an airloop availability manager turns the system off.
 - Add new enum for OA limiting factor.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by mjwitte** [comment]
This is a hopeful result for AirflowNetwork_MultiZone_SmallOffice_CoilHXAssistedDX. Will need to be sure the overall results are ok.
```
-   ** Severe  ** CheckWarmupConvergence: Loads Initialization, Zone="WEST ZONE" did not converge after 25 warmup days.
-   **   ~~~   ** See Warmup Convergence Information in .eio file for details.
-   **   ~~~   ** ...Environment(SizingPeriod)="MIAMI ANN CLG .4% CONDNS DB=>MWB"
-   **   ~~~   ** ..Max Temp Comparison = 2.46E-002 vs Temperature Convergence Tolerance=5.00E-003 - Fail Convergence
-   **   ~~~   ** ..Min Temp Comparison = 3.10E-003 vs Temperature Convergence Tolerance=5.00E-003 - Pass Convergence
-   **   ~~~   ** ..Max Heat Load Comparison = 8.4727E-003 vs Loads Convergence Tolerance=1.00E-003 - Fail Convergence
-   **   ~~~   ** ..Max Cool Load Comparison = 1.9900E-003 vs Loads Convergence Tolerance=1.00E-003 - Fail Convergence
```

 But we have new ones here in CentralChillerHeaterSystem_Simultaneous_Cooling_Heating. Win some and lose some.
```
+   ** Severe  ** CheckWarmupConvergence: Loads Initialization, Zone="SPACE3-1" did not converge after 25 warmup days.
+   **   ~~~   ** See Warmup Convergence Information in .eio file for details.
+   **   ~~~   ** ...Environment(SizingPeriod)="CHICAGO OHARE INTL AP ANN CLG 1% CONDNS DB=>MWB"
+   **   ~~~   ** ..Max Temp Comparison = 0.40 vs Temperature Convergence Tolerance=0.40 - Fail Convergence
+   **   ~~~   ** ..Min Temp Comparison = 3.05E-004 vs Temperature Convergence Tolerance=0.40 - Pass Convergence
+   **   ~~~   ** ..Max Heat Load Comparison = 0.0000 vs Loads Convergence Tolerance=4.00E-002 - Pass Convergence
+   **   ~~~   ** ..Max Cool Load Comparison = 2.0799E-005 vs Loads Convergence Tolerance=4.00E-002 - Pass Convergence
```

File: src/EnergyPlus/Furnaces.cc [hunk 1]
```diff
@@ -12099,6 +12099,7 @@ namespace Furnaces {
         if (state.dataFurnaces->Furnace(FurnaceNum).OpMode == ContFanCycCoil) state.dataHVACGlobal->OnOffFanPartLoadFraction = 1.0;
 
         auto &outNode = state.dataLoopNodes->Node(OutletNode);
+        AirMassFlow = outNode.MassFlowRate;
```
Comment (rraustad) [inline src/EnergyPlus/Furnaces.cc]: Up at 11520 is this line that can be deleted:

    AirMassFlow = state.dataLoopNodes->Node(InletNode).MassFlowRate;

CppCheck on this branch shows this:

    [src/EnergyPlus/Furnaces.cc:12102]:(style),[redundantAssignment],Variable 'AirMassFlow' is reassigned a value before the old one has been used.

File: src/EnergyPlus/Furnaces.cc [hunk 2]
```diff
@@ -9448,53 +9447,13 @@ namespace Furnaces {
...
-        Real64 QActual;           // heating coil load met or delivered
-        bool SuppHeatingCoilFlag; // .TRUE. if supplemental heating coil
-
-        FurnaceOutletNode = state.dataFurnaces->Furnace(FurnaceNum).FurnaceOutletNodeNum;
-        FurnaceInletNode = state.dataFurnaces->Furnace(FurnaceNum).FurnaceInletNodeNum;
-        CoolingCoilType_Num = state.dataFurnaces->Furnace(FurnaceNum).CoolingCoilType_Num;
-        HeatingCoilType_Num = state.dataFurnaces->Furnace(FurnaceNum).HeatingCoilType_Num;
-        WSHPRuntimeFrac = state.dataFurnaces->Furnace(FurnaceNum).WSHPRuntimeFrac;
-        CompPartLoadRatio = state.dataFurnaces->Furnace(FurnaceNum).CompPartLoadRatio;
+        auto &thisFurnace = state.dataFurnaces->Furnace(FurnaceNum);
```
Comment (Myoldmopar) [inline src/EnergyPlus/Furnaces.cc:9419]: References and variable scope cleanup :+1:

File: src/EnergyPlus/Furnaces.cc [hunk 3]
```diff
@@ -9562,322 +9517,304 @@ namespace Furnaces {
...
+            outletNode.Temp = Tout;
         }
 
         // If the fan runs continually do not allow coils to set OnOffFanPartLoadRatio.
         if (FanOpMode == ContFanCycCoil) state.dataHVACGlobal->OnOffFanPartLoadFraction = 1.0;
 
-        Real64 SensibleOutput(0.0); // sensible output rate, {W}
-        Real64 LatentOutput(0.0);   // latent output rate, {W}
-        Real64 TotalOutput(0.0);    // total output rate, {W}
+        Real64 SensibleOutput = 0.0; // sensible output rate, {W}
```
Comment (Myoldmopar) [inline src/EnergyPlus/Furnaces.cc:9767]: Assignment operator initialization is a fine change :+1:

File: src/EnergyPlus/Furnaces.hh [hunk 4]
```diff
@@ -360,21 +360,20 @@ namespace Furnaces {
...
+                           Real64 const CoolPartLoadRatio,                    // DX cooling coil part load ratio
+                           Real64 const HeatPartLoadRatio,                    // DX heating coil part load ratio (0 for other heating coil types)
+                           Real64 const HeatCoilLoad,                         // Heating coil load for gas heater
+                           Real64 const ReheatCoilLoad,                       // Reheating coil load for gas heater
+                           Real64 &SensibleLoadMet,   // Sensible cooling load met (furnace outlet with respect to control zone temp)
+                           Real64 &LatentLoadMet,     // Latent cooling load met (furnace outlet with respect to control zone humidity ratio)
+                           Real64 &OnOffAirFlowRatio, // Ratio of compressor ON mass flow rate to AVERAGE
+                           bool const HXUnitOn,       // flag to enable HX based on zone moisture load
+                           Real64 const CoolingHeatingPLRRat = 1.0 // cooling PLR to heating PLR ratio, used for cycling fan RH control
```
Comment (Myoldmopar) [inline src/EnergyPlus/Furnaces.hh:376]: Yes, a default value instead of an Optional :+1: :+1:

File: src/EnergyPlus/MixedAir.cc [hunk 5]
```diff
@@ -3341,7 +3334,7 @@ void OAControllerProps::CalcOAController(EnergyPlusData &state, int const AirLoo
         AirLoopCyclingFan = false;
     }
 
-    this->OALimitingFactor = limitFactorNone; // oa controller limiting factor
+    this->OALimitingFactor = static_cast<int>(OALimitFactor::None); // oa controller limiting factor
```
Comment (Myoldmopar) [inline src/EnergyPlus/MixedAir.cc:3337]: Hmm.  Hmmmm.  Is this right?  Is limiting factor intended to be a meaningful numeric value?  If not, why is `this->OALimitingFactor` not declared as the enum type?  I have a suspicion but let's see...

File: src/EnergyPlus/OutputReportTabular.cc [hunk 6]
```diff
@@ -6613,43 +6613,43 @@ void FillRemainingPredefinedEntries(EnergyPlusData &state)
         PreDefTableEntry(state,
                          state.dataOutRptPredefined->pdchOaTmFctNoLimit,
                          state.dataAirSystemsData->PrimaryAirSystems(iSys).Name,
-                         state.dataSysRpts->SysPreDefRep(iSys).TimeAtOALimit[MixedAir::limitFactorNone]);
+                         state.dataSysRpts->SysPreDefRep(iSys).TimeAtOALimit[static_cast<int>(MixedAir::OALimitFactor::None)]);
```
Comment (Myoldmopar) [inline src/EnergyPlus/OutputReportTabular.cc:6616]: So TimeAtOALimit is a std::array and you are doing an index lookup using the enum value.  This is fine of course given our current enum paradigm in E+.

**Review by Myoldmopar — COMMENTED** [review]
So I never saw why `OAControllerProps::OALimitingFactor` couldn't be declared as an `OALimitFactor` enum type.  If that could happen, it will clean up a chunk of those static cast assignments.  Static casting is fine when you need to use an enum as an index, but the raw assignments in there look a little out of place.

**Comment by Myoldmopar** [comment]
@mjwitte could you comment back on my review, and also I can't quite tell from your comment above whether you feel this is ready for primetime anyway.  Thanks!

**Comment by mjwitte** [comment]
> So I never saw why `OAControllerProps::OALimitingFactor` couldn't be declared as an `OALimitFactor` enum type. If that could happen, it will clean up a chunk of those static cast assignments. Static casting is fine when you need to use an enum as an index, but the raw assignments in there look a little out of place.

This gets reported as an output variable, and I'm pretty sure SetupOutputVariable won't accept an enum type. Would it be better to have 2 fields in the struct, the enum which would be used primarily, but then add an int flavor for reporting and just cast to it once somewhere?

And I set this PR back to draft, because there's no unit tests yet.

**Comment by Myoldmopar** [comment]
> This gets reported as an output variable,

That's what I expected as we've run into that in the past.  I don't think you need to worry about anything for now, leave it as-is.  IIRC, when we worked on this before, we thought we might be able to allow enums as output variables but we needed to modify the way we store the variable pointer in the output processor.  And this couldn't be done because of one assignment in the EMS manager.  I could be remembering it incorrectly, but either way, I don't think you need to change yours here, it's probably fine.

> And I set this PR back to draft, because there's no unit tests yet.

OK, it's off the list for today, thanks!

**Comment by mjwitte** [comment]
@rraustad Closing this one, superseded by #9947. There are some furnace refactors in this branch that may be worth picking up at some point, but don't hold up your cppcheck for Furnaces.cc

**Pull request closed without merging**